Add the Client-side Input Validators for Stream Keys and the Admin Password

dorj222 opened the pull request:
Currently, there are no client-side input validators for Stream Keys and the Admin Password. Setting a weak stream key or admin password could result in a scenario where the attacker could obtain access to the application controls. This PR implements the client-side input validators for stream keys and the admin password. These new input validators enforce the user to submit an input (a stream key or admin password) that must be a minimum of 8 characters and a maximum of 192 characters and must include at least 1 uppercase letter, at least one lowercase letter, at least 1 special character, and at least 1 number. [The Ant Design Form Validations](https://ant.design/components/form) have been used to implement these input validators. This PR fixes the issue #2528

1. The Input Validator for the Stream Keys:

<img width="831" alt="Screen Shot 2023-01-24 at 3 04 53 PM" src="https://user-images.githubusercontent.com/26856199/214315512-f08ea450-7666-4f26-9a05-b4728a1b03c3.png">

2. The Input Validator for the Admin Password:

<img width="785" alt="Screen Shot 2023-01-24 at 3 04 15 PM" src="https://user-images.githubusercontent.com/26856199/214315521-ff3f1f90-db21-4420-a4a7-d69f3b82b141.png">

dorj222 committed:
add the minimum stream key complexity rules on the client side

dorj222 committed:
add an admin password validator

dorj222 committed:
merge TextField and TextFieldAdmin components

dorj222 committed:
update Input Validators for Streak Keys and Admin Password

dorj222 committed:
fix a small regex typo

gabek commented:
Awesome, I'll check it out!

gabek commented on web/components/admin/config/server/StreamKeys.tsx:
```diff
@@ -49,16 +53,66 @@ const AddKeyForm = ({ setShowAddKeyForm, setFieldInConfigState, streamKeys, setE
+          },
+        ]}
+      >
+        <Input placeholder="def456" onChange={handleInputChange} />
```
We probably shouldn't set a simple placeholder if that's not an actual valid key.

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -53,14 +54,30 @@ export const TextField: FC<TextFieldProps> = ({
     // if an extra onChange handler was sent in as a prop, let's run that too.
     if (onChange) {
       const val = type === TEXTFIELD_TYPE_NUMBER ? e : e.target.value;
+      // console.log('val: ', val, 'fieldname: ', fieldName);
```
Nit: Console

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -53,14 +54,30 @@ export const TextField: FC<TextFieldProps> = ({
   };
 
+  useEffect(() => {
+    // console.log('value: ', value);
```
Nit: Console

gabek commented on web/components/admin/TextFieldWithSubmit.tsx:
```diff
@@ -134,15 +135,17 @@ export const TextFieldWithSubmit: FC<TextFieldWithSubmitProps> = ({
-            >
-              Update
-            </Button>
+            {fieldName !== 'adminPassword' && (
```
This feels like it shouldn't have a specific use case hard-coded in it.  Maybe if you're conditionally adding a button it can have a new property added to it that's like `hasSubmitButton` or something that can default to `true` but be set to `false` in the adminPassword case.

dorj222 replied:
I created a new `useState` isPwdInput boolean variable for this button and added a new `useEffect` function to check the `fieldName` and if it is the password input field would be set to true, else false. Please let me know if this change would resolve this code change request.

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -128,25 +145,87 @@ export const TextField: FC<TextFieldProps> = ({
-            disabled={disabled}
-            value={value as number | (readonly string[] & number)}
-          />
+      {fieldName !== 'adminPassword' ? (
```
Same comment about the hardcoded `adminPassword` check here, maybe a property can be added for the two different states.

dorj222 replied:
To fix this issue, I tried to create the `useState` boolean variable and `useEffect` function to listen to the variable `type` change for the input fields. However, when they were added the Regex Input Validators stopped working, and I could not find another way to resolve this issue. Currently, `type !== TEXTFIELD_TYPE_PASSWORD ` is being used to check if it is the password input field or not. Please feel free to suggest other possible solutions.

dorj222 committed:
code cleanup

dorj222 committed:
update Textfield and TextFieldWithSubmit

dorj222 committed:
Merge branch 'webv2' into fix/ImplementPasswordRules

dorj222 committed:
Prettified Code!

gabek commented:
Heads up: I changed the base of this PR from `webv2` to `develop` as all the feature work has been merged into `develop` and that's where changes will take place now.

dorj222 committed:
update the TextFieldWithSubmit component

gabek commented:
I know you're still working on this, so just ping me when you're ready for me to look at it again. Thanks so much for helping with this!

dorj222 committed:
Merge branch 'develop' into fix/ImplementPasswordRules

dorj222 commented:
Hi there @gabek! Would you please take a look at it again? :)

dorj222 committed:
correct the admin password endpoind API

dorj222 commented:
Actually, it turns out that changing the admin password endpoint API is causing some issues related to `const { enabled: federationEnabled } = federation;`. So, this error shows up whenever the admin password is successfully changed through the frontend admin. I will investigate this issue, and please feel free to suggest solutions that may fix this issue. 

Here is the error message: 
![Screen Shot 2023-02-05 at 4 40 43 PM](https://user-images.githubusercontent.com/26856199/216829440-0dffdae8-f84a-4c30-9fdd-659a43e66a8e.png)

gabek commented:
I just checked out your PR wasn't able to reproduce the error you mentioned above. Could you walk me through how to reproduce it?

dorj222 commented:
Sure! I changed the admin password through the frontend web application, refreshed the page after the password has been changed successfully, and then received the error that I showed. This error does not occur after adding a new stream key and only happens after the admin password change.

gabek commented:
What's going on is in development mode the password is always hard coded to `abc123`.  https://github.com/owncast/owncast/blob/develop/web/.env.development

In a production build it's fine. You can build production by running `./build/web/bundleWeb.sh` and accessing http://localhost:8080.

I know in this particular case it's super annoying, hopefully at some point in the future it can be improved on. Usually for development a hard coded password makes things faster to get things done, but in this case it's getting in the way.

<blockquote><img src="https://repository-images.githubusercontent.com/266445736/4f82fb80-a5b1-11ea-8f1c-69d316f861b5" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/owncast/owncast">owncast/.env.development at develop · owncast/owncast</a></strong></div><div>Take control over your live stream video by running it yourself.  Streaming + chat out of the box. - owncast/.env.development at develop · owncast/owncast</div></blockquote>
<blockquote><div><strong><a href="http://localhost:8080/">unfurl | built with Probot</a></strong></div></blockquote>

dorj222 commented:
ah, that really makes sense! thanks a lot for your prompt response! I think this PR is in a good shape. Would you please review it again?

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -53,14 +54,29 @@ export const TextField: FC<TextFieldProps> = ({
     if (onChange) {
       const val = type === TEXTFIELD_TYPE_NUMBER ? e : e.target.value;
+      setShowPwdButton(true);
+      if (fieldName === 'adminPassword' && regex.test(val)) {
```
Looks like there's still hardcoded references to an implementation of the text fields in the component itself with `adminPassword`. The generic components shouldn't have to know about specific instances of them.

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -53,14 +54,29 @@ export const TextField: FC<TextFieldProps> = ({
   };
 
+  useEffect(() => {
+    form.setFieldsValue({ adminPassword: value });
```
Another reference to adminPassword.

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -128,25 +144,88 @@ export const TextField: FC<TextFieldProps> = ({
+            <Form
+              name="basic"
+              form={form}
+              initialValues={{ adminPassword: value }}
```
Another reference to adminPassword.

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -128,25 +144,88 @@ export const TextField: FC<TextFieldProps> = ({
+              style={{ width: '100%' }}
+            >
+              <Form.Item
+                name="adminPassword"
```
Another reference to adminPassword.

gabek commented on web/components/admin/TextFieldWithSubmit.tsx:
```diff
@@ -66,6 +67,13 @@ export const TextFieldWithSubmit: FC<TextFieldWithSubmitProps> = ({
   }, [value]);
 
+  useEffect(() => {
+    if (fieldName === 'adminPassword') {
```
Another reference to adminPassword.

gabek commented:
Instead of hardcoding a concrete implementation of the component, could a boolean property be added to perform the logic that you're now doing by the string comparison? I know you mentioned above you were having issues removing that string comparison, but a generic component can't have internal logic based on one single use of it.

dorj222 committed:
refactor the Admin Password Input field and add a new boolean field for it

dorj222 committed:
refactor the Form Input field name from adminPassword to InputFieldPassword

dorj222 committed:
put password regex rules into config-constants.tsx

dorj222 commented:
Hi @gabek! Thank you so much for your feedback and suggestions. As you suggested, I added boolean variables inside the `TEXTFIELD_PROPS_ADMIN_PASSWORD`, `TEXTFIELD_PROPS_FFMPEG`, `TEXTFIELD_PROPS_WEB_PORT`, and `TEXTFIELD_PROPS_RTMP_PORT`, and these boolean variables are used to perform the logic. I rid of all the `fieldName === 'adminPassword'` comparisons. Also, I moved Regex constants into the `config-contstants` in order to make the code more readable and consistent. If it is possible, would please review it one more time again?

dorj222 committed:
regex constant typo fix

gabek commented:
Absolutely, I'm on it!

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -128,25 +150,66 @@ export const TextField: FC<TextFieldProps> = ({
+                  value={value as number | (readonly string[] & number)}
+                />
+              </Form.Item>
+              {showPwdButton && (
```
I see you added a submit button to `TextField`, but there's already a `TextFieldWithSubmit` that has a built-in submit button. It seems like these are now somewhat the same thing with them both showing a submit button on change. Sorry if I'm a little confused about needing to add a button to the one without a button.

dorj222 replied:
Hi, @gabek! The main reason why I add the separate submit button for the Admin Password field is that if we use the generic submit button of `TextFieldWithSubmit`, then the user would be able to submit the input even if their input is not satisfying the password complexity rules. Here in this post, I attached the screenshot example for this case. The separate submit button of `TextField` essentially prevents that problem and enforces the user to type a password that satisfies the password complexity rules. 

The screenshot example:
<img width="708" alt="Screen Shot 2023-02-21 at 3 22 51 PM" src="https://user-images.githubusercontent.com/26856199/220371875-60858fde-2f2b-4996-bcb5-63daa47c7ed0.png">

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -52,15 +55,30 @@ export const TextField: FC<TextFieldProps> = ({
   value,
+  isAdminPwdField,
 }) => {
+  const [hasPwdChanged, setHasPwdChanged] = useState(false);
```
Since `TextField` is used for all kinds of text input, not just passwords, could we use generic naming of things? Maybe `hasChanged` or `hasContentChanged` or something so it doesn't confuse somebody who's looking at the code later and wondering what a password has to do with it.

dorj222 replied:
I named them `hasPwdChanged` and `setHasPwdChanged` because they are only for the Admin Password input field and are password specific. However, I can change them to more generic names like `hasChanged` or `hasContentChanged`.  This `useState` makes the submit button for the Admin Password input field visible from the hidden state whenever the user types something inside the Admin Password input field. Also, we can make the submit button for the Admin Password always visible and remove this specific `useState`.

gabek commented on web/components/admin/TextField.tsx:
```diff
@@ -128,25 +150,66 @@ export const TextField: FC<TextFieldProps> = ({
-            disabled={disabled}
-            value={value as number | (readonly string[] & number)}
-          />
+      {!isAdminPwdField ? (
```
Similar comment, since these text fields aren't password specific. If what you're splitting up are fields with complexity requirements and fields without complexity requirements maybe the bool should be called `hasComplexityRequirements`.

gabek commented:
Sorry for having more questions and a few thoughts. Thank you for making those other changes, I think it's on the right path now with not putting implementation specific references in the generic component that's used everywhere in the admin. I just had a couple nits about calling things "password" since it's a generic component and can be used for any kind of input, and a question about the duplication of submit buttons between the two input components.

dorj222 commented:
Hi @gabek! I will work on these code change requests soon and will you know when it's ready. Thanks a lot for your feedback!

gabek commented:
Sounds great, thanks for the update!

dorj222 committed:
change the boolean variable isAdminPwdField to hasComplexityRequirements

gabek commented:
Hey @dorj222, I think we're good to get this in once the merge conflict is resolved.

dorj222 committed:
fix a merge conflict

dorj222 committed:
Merge branch 'develop' into fix/ImplementPasswordRules

dorj222 committed:
Prettified Code!

dorj222 commented:
Hi @gabek! The merge conflict has been resolved. Would you please review it again? 🙂

gabek commented:
Awesome, will do!

gabek commented:
Thanks so much for continuing to work on this, let's get it in! :1st_place_medal:

gabek commented:
Bad news, this PR is breaking the build and it needed to be reverted. I don't see build errors in the test above, but it's broken for me and others.

```
ReferenceError: Cannot access 'TEXTFIELD_TYPE_URL' before initialization
    at Object.xA (/home/gabek/src/webv2-merge/owncast/web/.next/server/chunks/6960.js:57:51)
    at 745 (/home/gabek/src/webv2-merge/owncast/web/.next/server/chunks/6960.js:1379:94)
    at __webpack_require__ (/home/gabek/src/webv2-merge/owncast/web/.next/server/webpack-runtime.js:25:42)
    at 9139 (/home/gabek/src/webv2-merge/owncast/web/.next/server/chunks/6960.js:68:81)
    at __webpack_require__ (/home/gabek/src/webv2-merge/owncast/web/.next/server/webpack-runtime.js:25:42)
    at 7990 (/home/gabek/src/webv2-merge/owncast/web/.next/server/pages/admin/config-federation.js:21:85)
    at __webpack_require__ (/home/gabek/src/webv2-merge/owncast/web/.next/server/webpack-runtime.js:25:42)
    at __webpack_exec__ (/home/gabek/src/webv2-merge/owncast/web/.next/server/pages/admin/config-federation.js:525:39)
    at /home/gabek/src/webv2-merge/owncast/web/.next/server/pages/admin/config-federation.js:526:100
    at __webpack_require__.X (/home/gabek/src/webv2-merge/owncast/web/.next/server/webpack-runtime.js:177:21)

> Build error occurred
Error: Failed to collect page data for /admin/config-federation
    at /home/gabek/src/webv2-merge/owncast/web/node_modules/next/dist/build/utils.js:991:15
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  type: 'Error'
}
```

gabek commented:
I went ahead and fixed the circular dependency import and the build is now successful. It's merged back in.
